src/System.Private.DataContractSerialization/src/System/Xml/XmlDictionaryAsyncCheckWriter.cs

```diff
@@ -0,0 +1,695 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace System.Xml
+{
+    internal class XmlDictionaryAsyncCheckWriter : XmlDictionaryWriter
+    {
+        private readonly XmlDictionaryWriter _coreWriter = null;
+        private Task _lastTask;
+
+        public XmlDictionaryAsyncCheckWriter(XmlDictionaryWriter writer)
+        {
+            _coreWriter = writer;
+        }
+
+        internal XmlDictionaryWriter CoreWriter
+        {
+            get
+            {
+                return _coreWriter;
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void CheckAsync()
+        {
+            if (_lastTask != null && !_lastTask.IsCompleted)
+            {
+                throw new InvalidOperationException(SR.XmlAsyncIsRunningException);
+            }
+        }
+
+        private Task SetLastTask(Task task)
+        {
+            _lastTask = task;
+            return task;
+        }
```

Comment:
Other XmlWriters will throw when async operations are called and if XmlWriterSettings.Async is `false`, should we do the same here?